### PR TITLE
Refactor tail

### PR DESCRIPTION
**Pull request by Joining7943**
I feel like the tail.rs file needs some refactoring to make it easier to implement new features. The file grew big over time, so I started with moving functions and structs around and moved them into new modules without changing much of the logic. The tests are all still passing, but there's much more potential for refactorings. Right now, I've spent just a few hours with it. I could dive deeper into the issue, however, I would appreciate to hear some feedback first and maybe you've got some additional ideas, since you've worked on tail much longer than I did.

EDIT:

Here's a quick overview of changes to `tail` in this step:

*  Extraction of smaller specialized modules from `tail.rs`. New modules are
   - paths: Containing logic around paths and inputs
   - args: Deal with argument parsing
   - follow/files: Logic around the paths followed by a `Watcher`
   - follow/watch: Contains the logic around following files and wraps a `notify::Watcher`
   - text: Small module to contain static strings used in every module
* Extraction of structs and methods in general
* Rewrite and compact the tail::uu_tail function
* Deal with stdin only when necessary, so when it is given on the commandline.
* Remove `mutable`s from `Settings` . Settings only store the initial values.
* Do not use PathBufs to store strings like "-"
* Simplify nested structures where possible
 
This PR is part of a larger refactoring, so there is still work that I address in a following PR.

### CODE REVIEW

**Comment by tertsdiepraam** [comment]
cc @jhscheer, who implemented #2695

File: build.rs [hunk 1]
```diff
@@ -70,6 +70,17 @@ pub fn main() {
...
+                phf_map.entry("tail", "(tail::uumain, tail::args::uu_app)");
+                tf.write_all(
+                    format!(
+                        "#[path=\"{dir}/test_tail.rs\"]\nmod test_tail;\n",
+                        dir = util_tests_dir,
+                    )
+                    .as_bytes(),
+                )
+                .unwrap();
+            }
```
Comment (tertsdiepraam) [inline build.rs:83]: Maybe it makes more sense to re-export `uu_app` from `tail`? I.e.

```rust
pub use args::uu_app;
```

I do like that you split the args into a separate file though!

**Review by tertsdiepraam — COMMENTED** [review]
Refactors are always welcome :) I can't look at this in depth now, but I have one small first comment. A quick scroll-through makes this look very promising!

**Comment by jhscheer** [comment]
Hi,
I'm with Terts, refactors are always welcome. However, there're a few things to watch out for.

> I feel like the tail.rs file needs some refactoring to make it easier to implement new features.

What features are you thinking of?

> The file grew big over time, so I started with moving functions and structs around and moved them into new modules without changing much of the logic. The tests are all still passing, but there's much more potential for refactorings.

Right now there're two CI monitored tests that don't pass in this refactor.
 A lot of `tail.rs` grew _organically_ to pass various tests from GNU's test suite. I transferred most of GNU's tests to our test frame work, but passing the test frame work does not mean passing GNU's tests. And to complicate things even more, not all of GNU's test suite tests that uu_tail passes are monitored by the CI (i.e. they pass on a local Linux, but not in the CI container). I mentioned this in tail's README.
To summarise, please regularly run GNU's tests after bigger changes to see if the relevant tests still pass.

> Right now, I've spent just a few hours with it.

Please wait for the other two open tail PRs to get merged before you do more refactoring, or you'll have to deal with merge conflicts.

**Comment by Joining7943** [comment]
> What features are you thinking of?

I had no specific features in mind. As far as I can see, uu_tail basically already provides all features gnu's tail does. I also meant to simplify fixing the `todo()`s and adding currently missing implementations for other platforms than linux, maybe also fixing some of the currently failing gnu tests.

> Please wait for the other two open tail PRs to get merged before you do more refactoring, or you'll have to deal with merge conflicts.

Sure, no problem. After 2 of the gnu tests failed, which originally passed, I thought about writing more tests before continuing the refactorings. Also, I would like to continue your work, incorporating gnu's test suite concerning tail into our own test framework. I'll take care as much as possible that the refactorings don't break existent code and tests.

**Comment by Joining7943** [comment]
Is breaking the `test_tail.rs` file in smaller related parts, like `test_tail/test_tail_follow.rs`, `test_tail/test_tail_args` etc. an option? It would simplify spotting missing tests. Naming the files `test_tail_...` won't break the `new_ucmd`, `util_name` macros. It's possible without changes in the build script, but then there would be a test_tail.rs file and a folder test_tail in the test directory. The other approach would involve a small addition to the build script which checks if there is a directory "test_tail" (in case of `tail`) in the test directory and if so adds the path `test_tail/test_tail.rs` to the test_modules.rs file. If there is no such directory the file `test_tail.rs` within the root of the test directory is added to test_modules like before. See also my last commit 4253aa46e24374a139317299fb267ab41d63605c as a quick solution.

**Comment by tertsdiepraam** [comment]
I think we could do that, but I don't quite understand the advantage you're talking about. Could you explain it a bit more?

**Comment by jhscheer** [comment]
> Is breaking the `test_tail.rs` file in smaller related parts, like `test_tail/test_tail_follow.rs`, `test_tail/test_tail_args` etc. an option? It would simplify spotting missing tests.

How?

If I'm only interested in a subset of tests I run, e.g.:
```bash
cargo test test_stdin --features "tail" --no-default-features
cargo test test_retry --features "tail" --no-default-features
```
etc.

**Comment by Joining7943** [comment]
The main benefit in my opinion is an organizational one. There are smaller files, which can be filled with more tests which are already partly explained and categorized by the file name. Then names for the tests themselves can be shorter or more specific to the problem, what in turn makes it easier to spot missing tests, if any. This won't break running only subsets of tests. In addition, one can use for example the filter `test_tail_follow` to run all tests in the file `test_tail/test_tail_follow.rs` and the test names must not necessarily include test_follow.

**Comment by jhscheer** [comment]
> Then names for the tests themselves can be shorter or more specific to the problem, what in turn makes it easier to spot missing tests, if any. This won't break running only subsets of tests. In addition, one can use for example the filter `test_tail_follow` to run all tests in the file `test_tail/test_tail_follow.rs` and the test names dont' have to necessarily include test_follow.

But that can already be done with modules, like what you did with `mod pipe_tests`.

**Comment by Joining7943** [comment]
Yes I did that and I didn't like it too much. I would have preferred moving the tests into an own file. However, I moved the tests into this mod to switch off the tests for windows in one go instead of annotating every single test. I don't see a reason keeping files big if there's no hard reason for it. Rust provides a nice way to split files into smaller units with ease and big files are just cumbersome to work with and most often unnecessary. If you think it's worth it, I'll move out the commit to an own pr and maybe fix the code duplication I introduce to the build file.

**Comment by tertsdiepraam** [comment]
My 2 cents: breaking up tests over files is unnecessary, because I usually just search for the test I need to look at, which is actually easier in one single file. For non-test code I agree that splitting is good, of course.

File: src/uu/tail/src/files.rs [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+//  * For the full copyright and license information, please view the LICENSE
+//  * file that was distributed with this source code.
+
+// spell-checker:ignore (ToDO) tailable seekable
+
+use crate::args::Settings;
+use crate::text;
+use core::option::Option;
+use core::option::Option::{None, Some};
+use core::result::Result::{Err, Ok};
```
Comment (tertsdiepraam) [inline src/uu/tail/src/files.rs:12]: All of these `Option`/`Result` things are in the prelude so don't need to be imported right?

File: src/uu/tail/src/files.rs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+        self.map.get(&Self::canonicalize_path(k)).unwrap()
+    }
+
+    /// Wrapper for HashMap::get_mut using Path::canonicalize
+    pub fn get_mut(&mut self, k: &Path) -> &mut PathData {
+        self.map.get_mut(&Self::canonicalize_path(k)).unwrap()
+    }
+
+    /// Canonicalize `path` if it is not already an absolute path
+    fn canonicalize_path(path: &Path) -> PathBuf {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/files.rs:70]: This feels like it should be a free standing function

**Review by tertsdiepraam — COMMENTED** [review]
A few small comments, but looks good overall. Of course we can only proceed with this of all the tests mentioned by @jhscheer still pass.

**Comment by Joining7943** [comment]
@tertsdiepraam Thanks for your comments and review. Current state is wip and I've waited to proceed with this pr until the two bigger prs around tail are merged before I make greater changes, especially to the logic. Right now, I just have to move things around in case of a merge conflict. I'm changing the code step by step working towards a final solution, executing the test suite after each step. At the moment, constructs may look useless, incomplete or illplaced.

**Comment by tertsdiepraam** [comment]
That sounds excellent! My comments were then a bit premature :)

**Comment by Joining7943** [comment]
No problem :) There's still a lot to do and I'll keep pushing intermediate results, so if you like you can see where the journey goes. I'll welcome any suggestions.

**Comment by tertsdiepraam** [comment]
Nice! Just remember to not make this PR too big. The bigger it gets the harder it is to review and there's not really a downside to having multiple successive PRs.

**Comment by Joining7943** [comment]
@tertsdiepraam I'm ready for review if you are. There's still a lot more that can be done and currently there's a last error on macos concerning `tail - < dir`. Like you suggested this'll be an intermediate pr. I'll fix the error, but this doesn't involve bigger changes. I'm also going to write a more exhaustive list of changes in the initial post of this pr.

File: src/uu/tail/src/tail.rs [hunk 4]
```diff
@@ -23,537 +23,68 @@ extern crate clap;
...
 }
 
-fn uu_tail(mut settings: Settings) -> UResult<()> {
-    let dash = PathBuf::from(text::DASH);
-
+fn uu_tail(settings: &Settings) -> UResult<()> {
     // Mimic GNU's tail for `tail -F` and exit immediately
-    if (settings.paths.is_empty() || settings.paths.contains(&dash)) && settings.follow_name() {
+    let mut input_service = InputService::from(settings);
+    let mut watcher_service = WatcherService::from(settings);
```
Comment (tertsdiepraam) [inline src/uu/tail/src/tail.rs:58]: I find the `InputService` and `WatcherService` to be a little bit vague. `Service` could mean basically anything.

File: src/uu/tail/src/tail.rs [hunk 5]
```diff
@@ -23,537 +23,68 @@ extern crate clap;
...
-        let metadata = path.metadata().ok();
-
-        if display_name.is_stdin() && !path.is_file() {
-            if settings.verbose {
-                files.print_header(&display_name, !first_header);
-                first_header = false;
+    for input in &input_service.inputs.clone() {
+        match input.kind() {
+            InputKind::File(path)
+                if cfg!(not(unix)) || cfg!(unix) && path != &PathBuf::from(text::DEV_STDIN) =>
```
Comment (tertsdiepraam) [inline src/uu/tail/src/tail.rs:73]:
```suggestion
                if cfg!(not(unix)) || path != &PathBuf::from(text::DEV_STDIN) =>
```

File: src/uu/tail/src/tail.rs [hunk 6]
```diff
@@ -567,742 +98,179 @@ fn uu_tail(mut settings: Settings) -> UResult<()> {
         the input file is not a FIFO, pipe, or regular file, it is unspecified whether or
         not the -f option shall be ignored.
         */
-        if files.no_files_remaining(&settings) {
-            if !files.only_stdin_remaining() {
-                show_error!("{}", text::NO_FILES_REMAINING);
-            }
-        } else if !(settings.stdin_is_pipe_or_fifo && settings.paths.len() == 1) {
-            follow(files, &settings, watcher_rx, orphans)?;
+        // TODO: can I do this check in watch::follow?
```
Comment (tertsdiepraam) [inline src/uu/tail/src/tail.rs:101]: There are a lot of TODO notes in here, which is fine, but because this is an open project, they need some context for everyone else who is reading the code.

File: src/uu/tail/src/args.rs [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+            }
+        }
+    } else {
+        return Err(ParseSizeError::ParseFailure(src.to_string()));
+    }
+
+    parse_size(size_string).map(|n| (n, starting_with))
+}
+
+// TODO: delete ?? not used anymore
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:314]: Please explain this todo. Is it about the function? If it's really not used it should indeed be removed.

File: src/uu/tail/src/args.rs [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+
+    /// Mode for lines delimited by delimiter as u8
+    Lines(Signum, u8),
+}
+
+impl FilterMode {
+    fn from(matches: &ArgMatches) -> UResult<Self> {
+        let zero_term = matches.contains_id(options::ZERO_TERM);
+        let mode = if let Some(arg) = matches.value_of(options::BYTES) {
+            match parse_num(arg) {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:67]: Can we make `parse_num` return `UResult<Signum>`?

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:314]: Tangentially related (and should not be fixed in this PR), but it feels like the `platform` module should handle this case correctly for Windows.

File: src/uu/tail/src/follow/files.rs [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        let k = Self::canonicalize_path(k);
+        if update_last {
+            self.last = Some(k.to_owned());
+        }
+        let _ = self.map.insert(k, v);
+    }
+
+    /// Wrapper for HashMap::remove using Path::canonicalize
+    pub fn remove(&mut self, k: &Path) -> PathData {
+        self.map.remove(&Self::canonicalize_path(k)).unwrap()
```
Comment (tertsdiepraam) [inline src/uu/tail/src/follow/files.rs:51]: Why is `unwrap` okay for these methods?

**Review by tertsdiepraam — COMMENTED** [review]
Cool! Some more small nits, but the big picture looks great I think. Have you checked the compared GNU tests locally as well with `main`, to ensure nothing broke that's not in the CI?

Also did we break these recently or have they never passed? In both cases well done!
```
Warning: Congrats! The gnu test tests/tail-2/follow-name is no longer failing!
Warning: Congrats! The gnu test tests/tail-2/retry is no longer failing!
Warning: Congrats! The gnu test tests/tail-2/wait is no longer failing!
```

File: src/uu/tail/src/follow/files.rs [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    /// Reload metadata from `path`, or `metadata`
+    pub fn update_metadata(&mut self, path: &Path, metadata: Option<Metadata>) {
+        self.get_mut(path).metadata = if metadata.is_some() {
+            metadata
+        } else {
+            path.metadata().ok()
+        };
+    }
+
+    /// Read new data from `path` and print it to stdout
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/follow/files.rs:138]: Sure. There's a lot of documenation missing in general. I can address this at the end of the next iteration.

File: src/uu/tail/src/follow/watch.rs [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+        if watcher_service.use_polling && settings.follow.is_some() {
+            // Consider all files to potentially have new content.
+            // This is a workaround because `Notify::PollWatcher`
+            // does not recognize the "renaming" of files.
+            paths = watcher_service.files.keys().cloned().collect::<Vec<_>>();
+        }
+
+        // main print loop
+        for path in &paths {
+            _read_some = watcher_service.files.tail_file(path, settings.verbose)?;
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/follow/watch.rs:583]: I'm pretty sure we can remove some or all of them, and this is certainly something we should address in the next pr.

File: src/uu/tail/src/paths.rs [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    fn is_seekable(&mut self, current_offset: u64) -> bool {
+        self.seek(SeekFrom::Current(0)).is_ok()
+            && self.seek(SeekFrom::End(0)).is_ok()
+            && self.seek(SeekFrom::Start(current_offset)).is_ok()
+    }
+}
+
+pub trait MetadataExtTail {
+    fn is_tailable(&self) -> bool;
+    fn got_truncated(&self, other: &Metadata) -> Result<bool, Box<(dyn UError)>>;
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/paths.rs:188]: Well seen!

File: src/uu/tail/src/tail.rs (hunk 6 above)
Comment (Joining7943) [reply src/uu/tail/src/tail.rs:101]: Sorry for that. I won't leave any TODOs when I'm finished, but this is just an intermediate pr, so I would like to keep them for now.

Comment (Joining7943) [reply src/uu/tail/src/tail.rs:101]: Ok I found a solution, so we can both be happy. I removed all TODOs in tail, so they don't end up in the main branch. I'll revert that commit later. No big deal.

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:314]: I've found that this function has no usages (anymore) within tail, so it's subject to deletion if I don't find any use for it.

File: src/uu/tail/src/args.rs (hunk 8 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:67]: Yep, I was a little bit torn here between keeping the function general or optimizing it. But, I'm on your side and also tend to the latter, here.

File: src/uu/tail/src/follow/files.rs (hunk 9 above)
Comment (Joining7943) [reply src/uu/tail/src/follow/files.rs:51]: I actually don't think it is and tend to return an Option or Result from these methods, but I need to dig deeper through the watch module to figure this out. These functions are heavily used. I can address this issue together with refactoring `watch` in the next pr.

File: src/uu/tail/src/tail.rs (hunk 4 above)
Comment (Joining7943) [reply src/uu/tail/src/tail.rs:58]: No problem. That's acutally one of the reasons, why I like using Service as a placeholder for such structures in the beginning of writing. I'll figure out something more meaningful and descriptive when I'm finishing this up and the purpose of these structures is clearified.

File: src/uu/tail/src/tail.rs [hunk 13]
```diff
@@ -427,14 +436,19 @@ fn bounded_tail(file: &mut File, settings: &Settings) {
...
+    // dbg!(&settings.mode);
+    match &settings.mode {
+        FilterMode::Lines(Signum::Negative(count), sep) => {
             let mut chunks = chunks::LinesChunkBuffer::new(*sep, *count);
             chunks.fill(reader)?;
             chunks.print(writer)?;
         }
-        (FilterMode::Lines(count, sep), true) => {
-            let mut num_skip = (*count).max(1) - 1;
+        FilterMode::Lines(Signum::PlusZero, _) | FilterMode::Lines(Signum::Positive(1), _) => {
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/tail.rs:447]: Thanks :)

File: src/uu/tail/src/follow/watch.rs [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+    pub fn follow_name_retry(&self) -> bool {
+        self.follow_name() && self.retry
+    }
+
+    fn init_files(&mut self, inputs: &VecDeque<Input>) -> UResult<()> {
+        if let Some(watcher_rx) = &mut self.watcher_rx {
+            for input in inputs {
+                match input.kind() {
+                    InputKind::Stdin => continue,
+                    InputKind::File(path) => {
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/follow/watch.rs:269]: Don't you think it adds some clearity keeping the match written out, here?

**Comment by Joining7943** [comment]
Thanks for this great review!

> Have you checked the compared GNU tests locally as well with main, to ensure nothing broke that's not in the CI?

Good, you're pointing that out. I didn't knew this about the CI, so I haven't run the tests locally for the last 30 commits or so. I'll run them tomorrow together with the fix for the failing test on macos.

> Also did we break these recently or have they never passed? In both cases well done!
Thanks. Imho they pass since this refactoring and cleanup, but I haven't picked out these tests intentionally, so I can't say for sure.

File: src/uu/tail/src/follow/watch.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/tail/src/follow/watch.rs:266]: An if clearly conveys that we're only using one arm in my opinion. I see what you mean though. If you think this is clearer then you can leave it like this.

File: src/uu/tail/src/tail.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/uu/tail/src/tail.rs:441]: I just realised you could use or-patterns here:
```
FilterLines::Lines(SigNum::PosZero | SigNum::Positive(1), _)
```

File: src/uu/tail/src/args.rs [hunk 15]
```diff
@@ -62,9 +130,10 @@ pub enum FollowMode {
     Name,
 }
 
+// TODO: use getters instead of pub access to fields to make Settings not changeable from the outside ?
```
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/args.rs]: agreed

File: src/uu/tail/src/follow/watch.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Joining7943) [reply src/uu/tail/src/follow/watch.rs:266]: I would like to keep it for now. It's just, that I'm not done with this part, and this little reminder helps me a little bit. I'll compact the match like you suggested in the next pr.

File: src/uu/tail/src/tail.rs (hunk 6 above)
Comment (jhscheer) [reply src/uu/tail/src/tail.rs]: It looks like you removed **all** TODOs, even those that were there before this refactor. This is very inconvenient.

File: src/uu/tail/src/tail.rs [hunk 16]
```diff
@@ -23,537 +23,66 @@ extern crate clap;
...
-    uu_tail(settings)
+    let settings = parse_args(args)?;
+    uu_tail(&settings)
 }
 
-fn uu_tail(mut settings: Settings) -> UResult<()> {
-    let dash = PathBuf::from(text::DASH);
-
+fn uu_tail(settings: &Settings) -> UResult<()> {
     // Mimic GNU's tail for `tail -F` and exit immediately
```
Comment (jhscheer) [inline src/uu/tail/src/tail.rs:56]:
```suggestion
```

File: src/uu/tail/src/tail.rs [hunk 17]
```diff
@@ -23,537 +23,66 @@ extern crate clap;
...
-    if (settings.paths.is_empty() || settings.paths.contains(&dash)) && settings.follow_name() {
+    let mut input_service = InputService::from(settings);
+    let mut watcher_service = WatcherService::from(settings);
+
```
Comment (jhscheer) [inline src/uu/tail/src/tail.rs:59]:
```suggestion
    // Mimic GNU's tail for `tail -F` and exit immediately
```

File: src/uu/tail/src/tail.rs [hunk 18]
```diff
@@ -1430,25 +380,33 @@ fn backwards_thru_file(file: &mut File, num_delimiters: u64, delimiter: u8) {
 /// being a nice performance win for very large files.
 fn bounded_tail(file: &mut File, settings: &Settings) {
     // Find the position in the file to start printing from.
-    match (&settings.mode, settings.beginning) {
-        (FilterMode::Lines(count, delimiter), false) => {
+    // dbg!("bounded");
+    // dbg!(&settings.mode);
```
Comment (jhscheer) [inline src/uu/tail/src/tail.rs:384]:
```suggestion
```

File: src/uu/tail/src/tail.rs [hunk 19]
```diff
@@ -1460,14 +418,19 @@ fn bounded_tail(file: &mut File, settings: &Settings) {
 fn unbounded_tail<T: Read>(reader: &mut BufReader<T>, settings: &Settings) -> UResult<()> {
     let stdout = stdout();
     let mut writer = BufWriter::new(stdout.lock());
-    match (&settings.mode, settings.beginning) {
-        (FilterMode::Lines(count, sep), false) => {
+    // dbg!("unbounded");
+    // dbg!(&settings.mode);
```
Comment (jhscheer) [inline src/uu/tail/src/tail.rs:422]:
```suggestion
```

**Review by jhscheer — COMMENTED** [review]
Except for the nit-picks, this looks very good to me.

File: src/uu/tail/src/tail.rs (hunk 6 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/tail.rs]: Yeah I wasn't suggesting to remove them, I just wanted them to be a bit more verbose so other people might pick them up more easily if they come across them in the code :)

**Comment by Joining7943** [comment]
Like I said this is an intermediate pr. @tertsdiepraam If there's nothing essentially wrong I would like to go on.

**Comment by tertsdiepraam** [comment]
I think we should merge this once this round of (simple) comments have been fixed so that we don't lose track of them. And then we can move to the next PR.

**Comment by Joining7943** [comment]
@tertsdiepraam I took care of every one of your comments, and then I proceeded. The other comments don't add any value, because I already addressed them on my new branch and are included in the follow up pr.

**Comment by tertsdiepraam** [comment]
Alright! Looking forward to the next one!

**Comment by jhscheer** [comment]
> The other comments don't add any value, because I already addressed them on my new branch and are included in the follow up pr.

Sorry what?

I did a review when this PR was marked ready for review.
Why do my comments not add any value? What does that even mean?

Where is this new branch?
Why did I spend time to address issues if they were already fixed some where?

How do you expect reviewers to keep track of issues they raise if the issues are not handled in the PR, but some where else?

@tertsdiepraam I think that merging a PR with open issues is very unusual.

**Comment by tertsdiepraam** [comment]
I interpreted not adding value as already fixed in a larger change, so in a sense outdated. I see refactoring as an ongoing change anyway, so I think it's fine in this case, but it is indeed unusual. The diff is getting huge and I'd rather review smaller changes from this point onward. So, @Joining7943, please keep it small from now on :)
